### PR TITLE
implementation of proposed pairs selection framework section a, b, c.

### PR DESCRIPTION
**Pull request by joelowj**
Hi @Jackal08, this is my humble attempt of implementing the amazing work "Enhancing a Pairs Trading strategy with the application of Machine Learning".

### CODE REVIEW

File: joel_ong/.gitignore [hunk 1]
```diff
@@ -0,0 +1,138 @@
+# Byte-compiled / optimized / DLL files
```
Comment (Jackal08) [inline joel_ong/.gitignore:1]: Quite a lot of extra items here but good that you added it.

File: joel_ong/dataset/constituents.csv [hunk 2]
```diff
@@ -0,0 +1,506 @@
+Symbol,Name,Sector
```
Comment (Jackal08) [inline joel_ong/dataset/constituents.csv:1]: It won't let me add comments to your notebook file since it is quite large. I will leave the comments for that file here.

File: joel_ong/dataset/constituents.csv [hunk 3]
```diff
@@ -0,0 +1,506 @@
...
+MMM,3M Company,Industrials
+AOS,A.O. Smith Corp,Industrials
```
Comment (Jackal08) [inline joel_ong/dataset/constituents.csv:3]: Missing the use of markdown with a clear abstract, intro, body and conclusion.

File: joel_ong/dataset/constituents.csv [hunk 4]
```diff
@@ -0,0 +1,506 @@
...
+ABT,Abbott Laboratories,Health Care
+ABBV,AbbVie Inc.,Health Care
```
Comment (Jackal08) [inline joel_ong/dataset/constituents.csv:5]: This notebook is ok. It shows how the functions are used and it gets the job done, although not in the detail that I was hoping for.

File: joel_ong/dataset/constituents.csv [hunk 5]
```diff
@@ -0,0 +1,506 @@
...
+ABMD,ABIOMED Inc,Health Care
+ACN,Accenture plc,Information Technology
```
Comment (Jackal08) [inline joel_ong/dataset/constituents.csv:7]: I like that you found 5 min data. Good initiative there.

File: joel_ong/dataset/constituents.csv [hunk 6]
```diff
@@ -0,0 +1,506 @@
...
+ATVI,Activision Blizzard,Communication Services
+ADBE,Adobe Inc.,Information Technology
```
Comment (Jackal08) [inline joel_ong/dataset/constituents.csv:9]: Nice tsne plot but you are missing axis labels and a title.

File: joel_ong/dataset/constituents.csv [hunk 7]
```diff
@@ -0,0 +1,506 @@
...
+AAP,Advance Auto Parts,Consumer Discretionary
+AMD,Advanced Micro Devices Inc,Information Technology
```
Comment (Jackal08) [inline joel_ong/dataset/constituents.csv:11]: Plotting the demeaned log prices of the various clusters wasn't as meaningful for me. There is a lot going on in a single plot. I agree that you can see the prices move similarly. Why did you demean the prices? Why not cumulative returns?

This was the first time that I saw demeaning in this context.

File: joel_ong/dataset/constituents.csv [hunk 8]
```diff
@@ -0,0 +1,506 @@
...
+AES,AES Corp,Utilities
+AFL,AFLAC Inc,Financials
+A,Agilent Technologies Inc,Health Care
```
Comment (Jackal08) [inline joel_ong/dataset/constituents.csv:14]: I would have liked to see some summary statistics of the pairs found in the data set you provided. A function that highlighted those in the same way the pandas .describe() function does it, but for stat arb.

File: joel_ong/dataset/constituents.csv [hunk 9]
```diff
@@ -0,0 +1,506 @@
...
+APD,Air Products & Chemicals Inc,Materials
+AKAM,Akamai Technologies Inc,Information Technology
```
Comment (Jackal08) [inline joel_ong/dataset/constituents.csv:16]: This notebook is a little sparse.

File: joel_ong/ml_pairs.py [hunk 10]
```diff
@@ -0,0 +1,374 @@
+#!/usr/bin/env python3
```
Comment (Jackal08) [inline joel_ong/ml_pairs.py:1]: What IDE did you use? Not PyCharm?

File: joel_ong/ml_pairs.py [hunk 11]
```diff
@@ -0,0 +1,374 @@
...
+# coding: utf-8
+# author: joelowj
+
+import logging
```
Comment (Jackal08) [inline joel_ong/ml_pairs.py:5]: Nice import order

File: joel_ong/ml_pairs.py [hunk 12]
```diff
@@ -0,0 +1,374 @@
...
+from statsmodels.tsa.stattools import coint
+
+from tqdm import tqdm
+from typing import List
+from typing import Tuple
+
+from utils import istarmap
+
+
+logging.config.fileConfig(fname="logger.conf", disable_existing_loggers=False)
```
Comment (Jackal08) [inline joel_ong/ml_pairs.py:32]: I like the use of logging.

File: joel_ong/ml_pairs.py [hunk 13]
```diff
@@ -0,0 +1,374 @@
...
+
+        This class implements Section III Pair Selection Framework, part
+            A. Dimensionality Reduction
+            B. Unsupervised Learning Clustering
+            C. Pairs Selection Criteria
+    """
+
+    def __init__(self, universe: pd.DataFrame, seed: int = 42) -> None:
+        """
+
```
Comment (Jackal08) [inline joel_ong/ml_pairs.py:49]: Missing docsting.

Comment (Jackal08) [reply joel_ong/ml_pairs.py:49]: Important because it explains the arguments to the user for this class. Without it, the useability is strained.

File: joel_ong/ml_pairs.py [hunk 14]
```diff
@@ -0,0 +1,374 @@
...
+            principal component, and then use the number of components that
+            explain a fixed percentage. A different approach is adopted. Because
+            an Unsupervised Learning algorithm is applied using these data features,
+            there should be a consideration for the data dimensionality.
+
+            Args:
+              top_k (int):
+                Determine number of PCA dimensions, upper bounded at 15.
+
+              norm_window (int):
```
Comment (Jackal08) [inline joel_ong/ml_pairs.py:74]: Clever idea.

File: joel_ong/ml_pairs.py [hunk 15]
```diff
@@ -0,0 +1,374 @@
...
+                Determine the rolling window if we were to apply rolling
+                normalization to the price series. If None, we will apply
+                normalization using the global mean and standard deviation.
+
+            Returns:
+
+        """
+        logger.info(f'Begining Part A. Dimensionality reduction.')
```
Comment (Jackal08) [inline joel_ong/ml_pairs.py:82]: Your code is a little clumped up together. I know others may like this but I prefer a standard that improves readability. For this reason, please use more whitespaces and treat your code in the same way that you would a paragraph, each one containing a specific idea.

In this case, its hard to tell when a new idea starts and an old one ends.

File: joel_ong/ml_pairs.py [hunk 16]
```diff
@@ -0,0 +1,374 @@
...
+        if norm_window:
+            universe_norm_return_df = (
+                universe_return_df -
+                universe_return_df.rolling(window=norm_window).mean()
+            ) / (universe_return_df.rolling(window=norm_window).std())
+            universe_norm_return_df.dropna(how='all', axis=0, inplace=True)
+        else:
+            scaler = StandardScaler()
+            universe_norm_return = scaler.fit_transform(universe_return_df)
+        pca = PCA(
```
Comment (Jackal08) [inline joel_ong/ml_pairs.py:103]: The style your coding in here is transferred from another language? Going forward I would change it to match our style. It's ok for now.

Comment (Jackal08) [reply joel_ong/ml_pairs.py:103]: The readability with the whitespaces after a brace is suffering.

File: joel_ong/ml_pairs.py [hunk 17]
```diff
@@ -0,0 +1,374 @@
...
+            universe_norm_return
+        )
+        self.features = pd.DataFrame(
+            pca.components_.T,
+            index=self._tickers
+        )
+
+    def plot_principle_components(self, figsize: tuple = (15, 10)) -> None:
+        if self.features is None:
+            raise ValueError(
```
Comment (Jackal08) [inline joel_ong/ml_pairs.py:116]: Nice

File: joel_ong/ml_pairs.py [hunk 18]
```diff
@@ -0,0 +1,374 @@
...
+            raise ValueError("""
+                The recommended PCA dimensions is lower bounded at 1 and
+                upper bounded at 15.
+            """)
+        universe_df = self.universe.copy(deep=True)
+        universe_return_df = universe_df.pct_change() \
+            .iloc[1:] \
+            .copy(deep=True)
+        universe_return_df.fillna(0, inplace=True)
+        if norm_window:
```
Comment (Jackal08) [inline joel_ong/ml_pairs.py:94]: Yes, good that you included the normalization for PCA.

File: joel_ong/ml_pairs.py [hunk 19]
```diff
@@ -0,0 +1,374 @@
...
+    def plot_principle_components(self, figsize: tuple = (15, 10)) -> None:
+        if self.features is None:
+            raise ValueError(
+                """
+                Please perform dimensionality reduction to generate feature
+                representation for each assets in the universe before running
+                this function.
+                """
+            )
+        sm = pd.plotting.scatter_matrix(
```
Comment (Jackal08) [inline joel_ong/ml_pairs.py:123]: Hmm, I wonder if this naming convention will raise errors under a PEP8 linter.

File: joel_ong/ml_pairs.py [hunk 20]
```diff
@@ -0,0 +1,374 @@
...
+            self.features, alpha=.2, figsize=figsize)
+        # Hide all ticks
+        [s.set_xticks(()) for s in sm.reshape(-1)]
+        [s.set_yticks(()) for s in sm.reshape(-1)]
+
+    def clustering(self, technique: str, params: dict) -> None:
+        """
+            Apply an appropriate clustering algorithm.
```
Comment (Jackal08) [inline joel_ong/ml_pairs.py:131]: The docstrings you are writing don't match our convention and when looking at it I realise that Pycharm wills in the format automatically. Maybe you are using a different IDE? Else have you maybe decided to purposefully use this convention? 

I am doubtful that sphinx documentation will be able to translate your arguments.

File: joel_ong/ml_pairs.py [hunk 21]
```diff
@@ -0,0 +1,374 @@
...
+                """
+            )
+        if technique == 'dbscan':
+            func_ = DBSCAN
+        elif technique == 'optics':
+            func_ = OPTICS
+        else:
+            raise ValueError(
+                f"The supported clustering techniques are dbscan and optics. Got {technique}.")
+        clustering = func_(**params).fit(self.features)
```
Comment (Jackal08) [inline joel_ong/ml_pairs.py:163]: The code suffers massively on readability here. 

Please use more:
* Inline comments
* Whitespace
* Don't whitespace after a brace `(`
* Use a linter
* Maybe increase max line length?

File: joel_ong/ml_pairs.py [hunk 22]
```diff
@@ -0,0 +1,374 @@
...
+        num_of_pairs = int(len(self._tickers) * (len(self._tickers) - 1) / 2)
+        logger.info(
+            f'Total number of candidate pairs (before clustering): {num_of_pairs}.')
+        num_of_pairs = int((self.valid_cluster_ids.value_counts(
+        ) * (self.valid_cluster_ids.value_counts() - 1)).sum() / 2)
+        logger.info(
+            f'Total number of candidate pairs (after clustering): {num_of_pairs}.')
+
+    def plot_cluster_members(self, figsize: tuple = (15, 10)) -> None:
+        if self.valid_cluster_ids is None:
```
Comment (Jackal08) [inline joel_ong/ml_pairs.py:180]: Missing doc strings...

File: joel_ong/ml_pairs.py [hunk 23]
```diff
@@ -0,0 +1,374 @@
...
+        )
+        plt.title('Cluster Members')
+        plt.xlabel('Number of Stock')
+        plt.ylabel('Cluster ID')
+        for i, v in enumerate(self.valid_cluster_ids.value_counts().tolist()):
+            ax.text(v + 1, i, str(v))
+        plt.show()
+
+    def plot_cluster_ts(
+        self,
```
Comment (Jackal08) [inline joel_ong/ml_pairs.py:197]: So hard to read this.

Comment (Jackal08) [reply joel_ong/ml_pairs.py:197]: Docstring

File: joel_ong/ml_pairs.py [hunk 24]
```diff
@@ -0,0 +1,374 @@
...
+            np.log(
+                self.universe[members] /
+                self.universe[members].mean())).plot(
+            figsize=figsize,
+            title=f'Demean Log Price for Stocks in Cluster {cluster_id}',
+            legend=False)
+        plt.show()
+
+    def plot_tsne(self, params: dict, figsize: tuple = (15, 10)) -> None:
+        params['random_state'] = self.seed
```
Comment (Jackal08) [inline joel_ong/ml_pairs.py:219]: Docstring

File: joel_ong/ml_pairs.py [hunk 25]
```diff
@@ -0,0 +1,374 @@
...
+        scatter = ax.scatter(
+            tsne[(self.cluster_ids.cluster_id == - 1).values.flatten(), 0],
+            tsne[(self.cluster_ids.cluster_id == - 1).values.flatten(), 1],
+            s=200,
+            alpha=.05,
+        )
+
+        plt.show()
+
+    def pairs_selection(
```
Comment (Jackal08) [inline joel_ong/ml_pairs.py:246]: You're not being consistent. Some functions are on a single line and others you revert to whitespaces after the brace.

File: joel_ong/ml_pairs.py [hunk 26]
```diff
@@ -0,0 +1,374 @@
...
+            self,
+            max_lag: int,
+            convenient_periods: int,
+            mp: bool) -> None:
+        """
```
Comment (Jackal08) [inline joel_ong/ml_pairs.py:251]: I stopped reading here. The readability could use work.

File: joel_ong/requirements.txt [hunk 27]
```diff
@@ -0,0 +1,7 @@
+matplotlib==3.3.2
```
Comment (Jackal08) [inline joel_ong/requirements.txt:1]: Good idea.

File: joel_ong/utils.py [hunk 28]
```diff
@@ -0,0 +1,104 @@
...
+# author: joelowj
+
+import multiprocessing.pool as mp
+import pandas as pd
+import yfinance as yf
+
+from tqdm import tqdm
+
+
+def download_dataset_yh(universe: list, filename: str = "universe") -> None:
```
Comment (Jackal08) [inline joel_ong/utils.py:12]: Docstrings missing. 

Your code will not pass PEP8 and linters.

**Review by Jackal08 — APPROVED** [review]
This submission was ok. 

Your notebook was lacking an abstract, intro, body, conclusion. It was very sparse. 

I can see that you're used to coding in a different language or that you're not familiar with PEP8 and linters. Your code suffers most on readability. 

The good news is that I think the readability and code style can be easily fixed. It looked like your technique was correct but it became very hard for me to follow. 

Will email you with a follow-up email. 

Please address all comments and if you are happy, then we will merge this PR into master.
